app/helpers/pipeline_runs_helper.rb

```diff
@@ -221,4 +224,30 @@ def check_for_user_error(failed_stage)
     end
     [nil, nil]
   end
+
+  # Return all pipeline runs that have succeeded for given samples
+  # Only check the first pipeline run.
+  # samples should be an ActiveRecord relation
+  # If strict mode is turned on, error out even if one pipeline run did not succeed.
+  # Note: Does NOT do access control checks.
+  def get_succeeded_pipeline_runs_for_samples(samples, strict = false)
+    # Gets the first pipeline runs for multiple samples in an efficient way.
+    created_dates = PipelineRun.select("sample_id, MAX(created_at) as created_at").where(sample_id: samples.pluck(:id)).group(:sample_id)
+    first_pipeline_runs = PipelineRun
+                          .select(:finalized, :id, :job_status)
+                          .where("(sample_id, created_at) IN (?)", created_dates)
```

Comment:
* You can remove `pipeline_runs` prefix actually! even more dry...
* Just add `.where(finalized: 1)` and you can remove the line below because pr is never going to be nil.